keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+when they serialize into their known version of a NodeResourceSlice. This tells
+them that they cannot handle the object because the API was extended.
+
+Compared to labels, attributes in this model have values of exactly one
+type. As described later on, these attributes get used in CEL expressions that
+the scheduler can use to select a specific resource for allocation on a node.
+
+To avoid repetition, common sets of
+attributes are defined once and referenced where needed. The example uses
+relatively simple sets, however, real-world use cases will include more information.
+
+The partitioning forms a tree. Nodes further down in that tree inherit the
+attributes of their parent and can overwrite those attributes where the values
+are different.
+
+In this example, "partitionable-gpu-1/half-and-quarters/second-half/quarter-0"
+indentifies one quarter of the second GPU in one particular configuration. It
+has "8Gi" of memory. Note that "memory" for this hardware is not something
+that can be sliced up and consumed arbitrarily. Instead, the partitioning
+defines how much if it is available, which makes the memory size an attribute
+of each individual sub-partition.
+
+The "partitionable-gpu-1" GPU can be used either with the "full-gpu", "quarters",
+or "half-and-quarters" partitioning scheme. When using "half-and-quarters", there's
+another choice between a "half" and "quarters" partitioning scheme for the second half. Once a
+partitioning scheme has been chosen to satisfy a claim, it cannot be changed
+until all claims bound to resources from that partition have been deallocated.
+
+At the moment, resource attribute names and their type are defined by
+vendors. Resource names with ".k8s.io" as a suffix are reserved for future use
+and standardization by Kubernetes. This could be used to describe topology
+across resources from different vendors, but this is out-of-scope for now.
+
+If a driver needs to reduce resource capacity, then there is a risk that a
+claim gets allocated using that capacity while the kubelet is updating a
+NodeResourceSlice. The implementations of semantic models must handle scenarios
+where more resources are allocated than available. The kubelet plugin of a DRA
+driver ***must*** double-check that the allocated resources are still available
+when NodePrepareResource is called. If not, the pod cannot start until the
+resource comes back. Treating this as a fatal error during pod admission would
+allow us to delete the pod and trying again with a new one.
+
+### Using semantic parameters
+
+The following is an example CRD which the developer of the
+`cards.dra.example.com` DRA driver might define as a valid claim parameters
+object for requesting access to its GPUs:
+
+```yaml
+kind: CardParameters
+apiVersion: dra.example.com/v1alpha1
+metadata:
+  name: my-parameters
+  namespace: user-namespace
+  uid: foobar-uid
+...
+spec:
+  count: 2
+  minimumRuntimeVersion: v12.0.0
```

Comment:
The way I have been thinking about it is the following...

The driver implementor provides vendor-specific CRDs which define the API for requesting resources governed by that driver. The driver controller then translates the "smeantics" encoded in these CRDs into CEL expressions for inclusion in the generic `ResourceClaimParameters` object.

An end user could always provide the `ResourceClaimParameters` object directly in their claim spec (without going through the vendor specific CRD), but trying to construct the proper boolean expressions to get at the resource they are interested in would likely be overly burdensome.

Reply:
Trying this on for size, so apologies if it doesn't work after I type it out :)

What if we cut it to the bone.  No vendor CRD at all.  A device is is a flattened collection of attributes, some of which are quantities.  A claim is a constraint expression (CEL) and a list of required minimum quantities.  That's it.  We'll provide enough CEL support to express whatever constraints we really need to express.  A vendor-specific process can take the minimums and quantize them (e.g. you asked for 768Mi, but we allocate in 1Gi units), but it's not a fundamental change - just quantized.

What doesn't work?

If you squint this is analogous to a pod which has node constraints (selector, etc) and resource requests.

Reply:
That's sort of where I started, too. You're missing the config options in the Class/Claim Parameter CRs, which are unrelated to scheduling but need to make their way down to the driver. But those don't need processing in this model, so maybe they could be put into the claim directly as a `RawExtension` (validation would be lost), or referenced from there directly as they are today (but permissions could get messy because without the vendor controller, scheduler needs to read them).

The indirection from ResourceClaim -> ClaimParameters also allows ClaimParameters to be re-used (think in a resourceClaimTemplate; though if it were embedded, it would just be copied in that case).

Perhaps a middle ground is to keep the basic design we have, but just use a very simple "semantic model" that captures a list of attributes (for selection) and a list of Quantity capacities (for allocation). We can debate whether selection pops up to the ResourceClaim and/or ResourceClaimParameters during implementation or 1.31.

Reply:
> CEL is for humans.

In the current iteration of the KEP, users might be writing those expressions. All they need are the necessary RBAC permissions to create ResourceClaimParameters objects.

In some prior revision I listed allowed API resources in the resource class, so the choice would have been by class and thus driver. That's now just the bare minimum (a boolean) which indicates that semantic parameters are to be used, regardless of the parameter type.

>  No vendor CRD at all.

My take on this: I expect that to be a very user-*un*friendly API which will be harder to understand, document, and validate. I could be wrong, but that's my gut feeling.

Let's try it out and come back to this once we have some more realistic examples?

If want to drop support for vendor CRDs, all we need to do is remove the `generatedFrom` and thus the indirection.

> What doesn't work?

Config parameters, but @johnbelamaric already mentioned those.

Reply:
Plumbing config params in seems simple to me, because (as John said) "those don't need processing".

> The indirection from ResourceClaim -> ClaimParameters also allows ClaimParameters to be re-used (think in a resourceClaimTemplate

I am not convinced this justifies its own weight.  If we reduce reduce reduce, then the "claim parameters" are simple enough that copying is better?

IOW: can the WHOLE problem reduce (more or less) to https://go.dev/play/p/gdnT0xef7iD ?

Reply:
I think we need to punt on this discussion for now.

I feel just as strongly *against* exposing CEL to the end-user as you do *for* exposing it. As a device vendor, I want to be in control of the API I present to my users (so that it matches their experience in other environments) and forcing them to write a complex CEL expression when I can hide it behind a single field in my custom API, doesn't feel like good UX to me. It also doesn't allow me to pass any configuration parameters through the claim (which should be considered a strict requirement).

Your proposal also doesn't seem to cover asking for multiple instances of a like-set of resources (i.e. give me 4 GPUs attached to this claim). This is necessary, for example, to run an MPS server that governs all GPUs of the claim and then share them across multiple containers / pods (which also requires custom configuration parameters in any case).

Reply:
I agree we should punt on this question. We can address as we learn more.

Reply:
The good (or bad, depending on POV) news is that CEL is already part of the API in several places and growing.  

> As a device vendor, I want to be in control of the API I present to my users

You are in control - you decide what params they can evaluate against.  

> forcing them to write a complex CEL expression when I can hide it behind a single field in my custom API

I'll ask for a concrete example here.

> doesn't allow me to pass any configuration parameters through the claim

I'm not seeing that conclusion.  Or rather, not _necessarily_.  Params defined in the vendor-centric class-params can be passed, and if we really think there are per-claim params to pass from a user (workload owner) into the driver which are not quantities requested, then passing vendor-centric claim-params is a logical step.  But my YAGNI alarms are firing. 

> Your proposal also doesn't seem to cover asking for multiple instances of a like-set of resources

https://go.dev/play/p/dZyI8qw0ufJ

I recognize that it seems like I have sort of jumped the rails here, but less really is more.    We have a complexity budget and I think we are over it right now.

Reply:
One further - the class can have constraints, too.

https://go.dev/play/p/JJmgKTUjzxV